### DISCUSSION TITLE
[Question] How to build autocomplete feature using tree-sitter ?

### DISCUSSION THREAD

**Discussion by KShivendu** [tree-sitter/tree-sitter#1149] [Q&A] [answered]
Hi,
I'm working on a project which is about building a custom search query language and keeping it accessible. This custom search query language has to be translated to Elasticsearch DSL and while writing the system should have autocomplete feature to guide the users. I could find some similar projects using ANTLR but I would prefer to build it using tree-sitter. 

I'm a beginner when it comes to building parsers and autocomplete features. So forgive me if I missed something obvious. Please comment below if you have any idea/resources about how to implement it using tree-sitter :)

Thank you

**Comment by razzeee** [comment, marked as answer]
Well, you can use tree-sitter to access the tokens in your code. But you will need to handle which completions are valid at any given time. So this is only slightly connected to tree-sitter in the end.

I basically do it like this:
1. Parse all my files with tree-sitter
2. Collect some more data like imports and how files relate to each other
3. Access the collected data to output completions via the language server protocol

**Reply by KShivendu** [reply]
Hey @Razzeee thank you very much. Actually, I missed an important point, my search query language will search for repositories from https://archive.softwareheritage.org The language would look like this: 
- `lang:TypeScript useState OR useMemo`
-  `repo:sourcegraph/sourcegraph type:diff after:"1 week ago"` 
- `repo:^github\.com/sourcegraph/sourcegraph$@3.17 CONTAINER_ID`

These examples were from https://sourcegraph.com/search which has similar autocomplete features for its query language. 

Also, after parsing these queries, we have to convert them into Elasticsearch queries (a JSON-like grammar) that will follow our Elasticsearch schema(mappings)

**Comment by maxbrunsfeld** [comment]
I moved this from Issues into Discussions. We're trying to use Discussions for these types of questions because they don't represent work items for the maintainers, and they never need to be closed out.
